### PR TITLE
Added Feature/config groups

### PR DESCRIPTION
**Pull request by Sunny-NEC**
Added code for iot/configGroups (https://github.com/telefonicaid/iotagent-node-lib/issues/752)

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -13,3 +13,4 @@
 - FIX: Remove preprocess stripping of explicitAttrs (#1151)
 - FIX: NGSI-LD commands are invalid. (#1185)
 - Upgrade logops dep from 2.1.0 to 2.1.2 due to colors dependency corruption
+dd /iot/configGroups API endpoints (as equivalent to /iot/services) (#752)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:16]: `dd` seems to be a typo

**Comment by fgalan** [comment]
Please solve the conflict in CHANGES_NEXT_RELEASE so unit tests can automatically run in your PR.

**Comment by fgalan** [comment]
It seems this is the fifthincarnation of the cgroups PR:

* First: #775
* Second: #780
* Third: #922
* Fourth: #954
* Fifth (this one): #1198 

Just in case it can be useful (mainly for taking into account all the potential feedback spreaded over them)

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (mapedraza) [inline CHANGES_NEXT_RELEASE:16]:
```suggestion
Add /iot/configGroups API endpoints (as equivalent to /iot/services) (#752)
```

File: doc/api.md [hunk 2]
```diff
@@ -203,47 +230,32 @@ outgoing requests).
...
+| `expressionLanguage`  | `expresionLanguage`  | optional boolean value, to set expression language used to compute expressions, possible values are: legacy or jexl. When not set or wrongly set, legacy is used as default value.                                                                                                                                          |
+| `explicitAttrs`       | `explicitAttrs`      | Boolean value to support selective ignore of measures for device so that IOTA doesn’t progress. If not specified default is `false`.                                                                                                                                                                                        | `true/false`                                  |
+| `ngsiVersion`         | `ngsiVersion`        | optional string value used in mixed mode to switch between **NGSI-v2** and **NGSI-LD** payloads. The default is `v2`. When not running in mixed mode, this field is ignored.                                                                                                                                                | `v2/ld`                                       |
```
Comment (mapedraza) [inline doc/api.md:258]: This change is not included in PR https://github.com/telefonicaid/iotagent-node-lib/pull/954

File: lib/services/northBound/northboundServer.js [hunk 3]
```diff
@@ -117,7 +117,6 @@ exports.setNotificationHandler = intoTrans(context, contextServer.setNotificatio
 exports.setConfigurationHandler = intoTrans(context, groupProvisioning.setConfigurationHandler);
 exports.setRemoveConfigurationHandler = intoTrans(context, groupProvisioning.setRemoveConfigurationHandler);
 exports.setProvisioningHandler = intoTrans(context, deviceProvisioning.setProvisioningHandler);
-exports.setUpdatingHandler = intoTrans(context, deviceUpdating.setUpdatingHandler);
```
Comment (mapedraza) [inline lib/services/northBound/northboundServer.js:120]: This change is not included in PR https://github.com/telefonicaid/iotagent-node-lib/pull/954

Could you elaborate why is included here?

File: test/unit/general/deviceService-test.js [hunk 4]
```diff
@@ -25,9 +25,9 @@
 
 const iotAgentLib = require('../../../lib/fiware-iotagent-lib');
 const utils = require('../../tools/utils');
-const request = utils.request;
```
Comment (mapedraza) [inline test/unit/general/deviceService-test.js:28]: This change is not included in PR https://github.com/telefonicaid/iotagent-node-lib/pull/954

Could you elaborate why is included here?

File: test/unit/general/deviceService-test.js [hunk 5]
```diff
@@ -165,10 +193,10 @@ let contextBrokerMock;
 let iotamMock;
 
 /* jshint camelcase: false */
-describe('NGSI-v2 - Device Service: utils', function () {
+describe('NGSI-v1 - Device Service: utils', function () {
```
Comment (mapedraza) [inline test/unit/general/deviceService-test.js:196]: Why you changed it to V1?

File: test/unit/general/deviceService-test.js [hunk 6]
```diff
@@ -212,19 +249,69 @@ describe('NGSI-v2 - Device Service: utils', function () {
...
                 .matchHeader('fiware-servicepath', '/testingPath')
-                .post('/v2/registrations')
-                .reply(201, null, { Location: '/v2/registrations/6319a7f5254b05844116584d' });
+                .post('/NGSI9/registerContext')
+                .reply(
+                    200,
+                    utils.readExampleFile(
+                        './test/unit/examples/contextAvailabilityResponses/registerProvisionedDeviceSuccess.json'
+                    )
+                );
```
Comment (mapedraza) [inline test/unit/general/deviceService-test.js:305]: NGSI V1 is no longer supported, Why you included those NGSIv1 tests in NGSIv1 the codebase?

File: test/unit/general/deviceService-test.js [hunk 7]
```diff
@@ -212,19 +249,69 @@ describe('NGSI-v2 - Device Service: utils', function () {
...
             contextBrokerMock
                 .matchHeader('fiware-service', 'testservice')
                 .matchHeader('fiware-servicepath', '/testingPath')
-                .post('/v2/entities?options=upsert')
-                .reply(204);
+                .post('/v1/updateContext')
+                .reply(
+                    200,
+                    utils.readExampleFile('./test/unit/examples/contextResponses/createProvisionedDeviceSuccess.json')
+                );
```
Comment (mapedraza) [inline test/unit/general/deviceService-test.js:314]: NGSI V1 is no longer supported, Why you included those NGSIv1 tests in NGSIv1 the codebase?

File: test/unit/ngsiv2/general/deviceService-test.js [hunk 8]
```diff
@@ -193,12 +221,42 @@ describe('NGSI-v2 - Device Service: utils', function () {
...
-                    done();
-                }
-            );
+            async.series([request.bind(request, groupCreation), request.bind(request, deviceCreation)], function (
+                error,
+                results
+            ) {
+                done();
+            });
+        });
```
Comment (mapedraza) [inline test/unit/ngsiv2/general/deviceService-test.js:230]: Those changes were not included in PR #954. Could you elaborate why are included here?

File: test/unit/ngsiv2/general/deviceService-test.js [hunk 9]
```diff
@@ -223,7 +314,7 @@ describe('NGSI-v2 - Device Service: utils', function () {
                 .post('/v2/entities?options=upsert')
                 .reply(204);
 
-            async.series([utils.request.bind(utils.request, groupCreation)], function (error, results) {
+            async.series([request.bind(request, configGroupCreation)], function (error, results) {
```
Comment (mapedraza) [inline test/unit/ngsiv2/general/deviceService-test.js:317]: Those changes were not included in PR #954. Could you elaborate why are included here?

File: test/unit/mongodb/mongodb-configGroup-registry-test.js [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+ * Modified by: Daniel Calvo - ATOS Research & Innovation
+ */
+
+/* eslint-disable no-unused-vars */
+
+const iotAgentLib = require('../../../lib/fiware-iotagent-lib');
+const _ = require('underscore');
+const utils = require('../../tools/utils');
+const async = require('async');
+const request = require('request');
```
[the remark this answers is not in the document]
Comment (Sunny-NEC) [reply test/unit/mongodb/mongodb-configGroup-registry-test.js]: https://github.com/telefonicaid/iotagent-node-lib/pull/1198/commits/be9e2b8908da9bbc265c3ca529b617a4c787c0ef
@mapedraza please review this commit . I did all the required changes.

**Comment by mapedraza** [comment]
overpassed by https://github.com/telefonicaid/iotagent-node-lib/pull/1375#issuecomment-1736903343

**Pull request closed without merging**